{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "98f19562",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import dask.dataframe as dd\n",
    "from dask.distributed import Client\n",
    "import pandas as pd\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "\n",
    "pd_data_dir = \"/home/djl34/lab_pd/data\"\n",
    "aso_data_dir = \"/home/djl34/lab_pd/aso/data\"\n",
    "KL_data_dir = \"/home/djl34/lab_pd/kl/data\"\n",
    "scratch_dir = \"/n/scratch3/users/d/djl34\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "717f180c",
   "metadata": {},
   "source": [
    "## get a list of genes that are AD and lof"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "54d4404d",
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = aso_data_dir + \"/Clingen-Gene-Disease-Summary-2023-06-24.csv\"\n",
    "df = pd.read_csv(filename)\n",
    "\n",
    "filename = aso_data_dir + \"/Clingen-Dosage-Sensitivity-2023-06-25.csv\"\n",
    "df_ds = pd.read_csv(filename)\n",
    "df_ds_haploinsufficient = df_ds[df_ds[\"HAPLOINSUFFICIENCY\"].isin(['Sufficient Evidence for Haploinsufficiency', 'Emerging Evidence for Haploinsufficiency'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "faa76d26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Gene Associated with Autosomal Recessive Phenotype',\n",
       "       'Sufficient Evidence for Haploinsufficiency',\n",
       "       'No Evidence for Haploinsufficiency',\n",
       "       'Little Evidence for Haploinsufficiency',\n",
       "       'Emerging Evidence for Haploinsufficiency',\n",
       "       'Dosage Sensitivity Unlikely'], dtype=object)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ds[\"HAPLOINSUFFICIENCY\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e49176dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Disputed', 'Definitive', 'Moderate', 'Limited',\n",
       "       'No Known Disease Relationship', 'Strong', 'Refuted'], dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"CLASSIFICATION\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9006b4e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_AD = df[df[\"MOI\"] == \"AD\"]\n",
    "\n",
    "df_AD_strong = df_AD[df_AD[\"CLASSIFICATION\"].isin([\"Strong\", \"Definitive\"])]\n",
    "\n",
    "df_AD_strong_lof = df_AD_strong.merge(df_ds_haploinsufficient, on = \"GENE SYMBOL\", how = \"inner\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9ea76676",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list = list(df_AD_strong_lof[\"GENE SYMBOL\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "29eb5e7b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "191"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd6d1614",
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = pd_data_dir + \"/biomart/ENSG_Genename_syn.tsv\"\n",
    "df_ensg = pd.read_csv(filename, sep = \"\\t\")\n",
    "df_ensg = df_ensg.rename({\"Gene name\": \"GENE SYMBOL\", \"Gene stable ID\" : \"Gene\"}, axis = 1)\n",
    "df_ensg = df_ensg[[\"GENE SYMBOL\", \"Gene\", \"Chromosome/scaffold name\"]].drop_duplicates()\n",
    "df_ensg = df_ensg[df_ensg[\"Chromosome/scaffold name\"].str.contains(\"CHR\") == False]\n",
    "df_ensg = df_ensg.rename({\"Gene name\": \"GENE SYMBOL\", \"Gene stable ID\" : \"Gene\"}, axis = 1)\n",
    "\n",
    "df_AD_strong_lof_ensg = df_AD_strong_lof.merge(df_ensg, on = \"GENE SYMBOL\", how = \"left\")\n",
    "\n",
    "ensg_gene_list = list(df_AD_strong_lof_ensg[\"Gene\"].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e18b8b7",
   "metadata": {},
   "source": [
    "## combine with lof data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "92f00f56",
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = aso_data_dir + \"/fake_transcript_variants_v4_lof_HC_filtered_unint.tsv\"\n",
    "df_lof = pd.read_csv(filename, sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e7cfb8ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Gene', 'Feature', 'Chrom', 'Pos', 'Allele_ref', 'Allele',\n",
       "       'Consequence', 'UNIPROT_ISOFORM', 'gnomAD_AF', 'gnomAD_NFE_AF',\n",
       "       'LoF_flags', 'CANONICAL', 'mu_nonhyper', 'mu_hypermut', 'mu_quality',\n",
       "       'context', 'mean', 'median', 'over_1', 'over_5', 'over_10', 'over_15',\n",
       "       'over_20', 'over_25', 'over_30', 'over_50', 'over_100', 'filter', 'AN',\n",
       "       'AN_nfe', 'AC', 'AC_nfe', 'controls_AN', 'controls_AC', 'non_cancer_AN',\n",
       "       'non_cancer_AC', 'BaseQRankSum', 'ClippingRankSum', 'DP', 'FS',\n",
       "       'InbreedingCoeff', 'MQ', 'MQRankSum', 'QD', 'ReadPosRankSum', 'SOR',\n",
       "       'VQSLOD', 'VQSR_NEGATIVE_TRAIN_SITE', 'VQSR_POSITIVE_TRAIN_SITE',\n",
       "       'VQSR_culprit', 'allele_type', 'dp_hist_all_n_larger',\n",
       "       'dp_hist_alt_n_larger', 'has_star', 'lcr', 'n_alt_alleles', 'pab_max',\n",
       "       'rf_label', 'rf_negative_label', 'rf_positive_label',\n",
       "       'rf_tp_probability_label', 'interp_dist', 'site_in_genome_bottle',\n",
       "       'ref_alt', 'Carlson_rate', 'tri', 'methylation_level', 'gnomAD_rate',\n",
       "       'mu', 'AC_unint', 'AC_nfe_unint'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_lof.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bdce6b39",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_lof = df_lof[[\"Gene\", \"Chrom\", \"Pos\", \"Allele_ref\", \"Allele\", \"Consequence\", \"LoF_flags\", \"filter\", \"mu\", \"AN\", \"AC_unint\", \"AN_nfe\", \"AC_nfe_unint\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0529d871",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_AD_strong_lof_ensg = df_AD_strong_lof.merge(df_ensg, on = \"GENE SYMBOL\", how = \"left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "338c6d47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['12', '20', '1', '4', '16', '5', '2', '6', '18', '11', '7', '10',\n",
       "       '3', '17', '13', '9', '15', '8', '14', '22', '21', '19'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_AD_strong_lof_ensg[\"Chromosome/scaffold name\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3ef03414",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_AD_strong_lof_ensg = df_AD_strong_lof_ensg[[\"GENE SYMBOL\", \"Gene\", \"DISEASE LABEL\", \"MOI\", \"CLASSIFICATION\", \"GCEP\", \"HAPLOINSUFFICIENCY\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2d9bc86f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE SYMBOL</th>\n",
       "      <th>Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACVRL1</td>\n",
       "      <td>ENSG00000139567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADNP</td>\n",
       "      <td>ENSG00000101126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AHDC1</td>\n",
       "      <td>ENSG00000126705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ANK2</td>\n",
       "      <td>ENSG00000145362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ANKRD11</td>\n",
       "      <td>ENSG00000167522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>VHL</td>\n",
       "      <td>ENSG00000134086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>WAC</td>\n",
       "      <td>ENSG00000095787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>WT1</td>\n",
       "      <td>ENSG00000184937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>ZEB2</td>\n",
       "      <td>ENSG00000169554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>ZNF462</td>\n",
       "      <td>ENSG00000148143</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>191 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    GENE SYMBOL             Gene\n",
       "0        ACVRL1  ENSG00000139567\n",
       "1          ADNP  ENSG00000101126\n",
       "2         AHDC1  ENSG00000126705\n",
       "3          ANK2  ENSG00000145362\n",
       "4       ANKRD11  ENSG00000167522\n",
       "..          ...              ...\n",
       "217         VHL  ENSG00000134086\n",
       "218         WAC  ENSG00000095787\n",
       "219         WT1  ENSG00000184937\n",
       "221        ZEB2  ENSG00000169554\n",
       "222      ZNF462  ENSG00000148143\n",
       "\n",
       "[191 rows x 2 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_AD_strong_lof_ensg[[\"GENE SYMBOL\", \"Gene\"]].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "15fceafd",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_AD_strong_lof_ensg[[\"GENE SYMBOL\", \"DISEASE LABEL\", \"HAPLOINSUFFICIENCY\"]].to_csv(\"../data/genes_AD_lof_strong.tsv\",\n",
    "                                                                                    sep = \"\\t\", index = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "75fbafc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_AD_strong_lof_ensg_variants = df_AD_strong_lof_ensg[[\"GENE SYMBOL\", \"Gene\"]].drop_duplicates().merge(df_lof, on = \"Gene\", how = \"left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c396c381",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_AD_strong_lof_ensg_variants.to_csv(\"../data/variants_AD_lof_strong.tsv\", sep = \"\\t\", index = None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9701ba15",
   "metadata": {},
   "source": [
    "## filter by splice acceptor variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "ece14d81",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_13654/2367571652.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_AD_strong_lof_ensg_variants[\"Chrom\"] = df_AD_strong_lof_ensg_variants[\"Chrom\"].astype(int)\n"
     ]
    }
   ],
   "source": [
    "df_AD_strong_lof_ensg_variants = df_AD_strong_lof_ensg_variants[df_AD_strong_lof_ensg_variants[\"Chrom\"].isna() == False]\n",
    "df_AD_strong_lof_ensg_variants[\"Chrom\"] = df_AD_strong_lof_ensg_variants[\"Chrom\"].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "75ab50e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_variants = df_AD_strong_lof_ensg_variants[df_AD_strong_lof_ensg_variants[\"Chrom\"].isna() == False]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "388dc371",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_variants = df_AD_strong_lof_ensg_variants[df_AD_strong_lof_ensg_variants[\"Chrom\"] == 22]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "c13106f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([12., 20.,  1.,  4., 16.,  5.,  2.,  6., 18., 11.,  7., 10.,  3.,\n",
       "       17., 13.,  9., 15.,  8., 14., 22., 21., 19., nan])"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_AD_strong_lof_ensg_variants[\"Chrom\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1adb9885",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['stop_gained', 'splice_acceptor_variant', 'splice_donor_variant',\n",
       "       'stop_gained&splice_region_variant'], dtype=object)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_variants[\"Consequence\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "0bf761fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_variants = df_variants[(df_variants[\"Consequence\"].str.contains(\"splice_acceptor_variant\")) | (df_variants[\"Consequence\"].str.contains(\"splice_donor_variant\"))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "41434af9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE SYMBOL</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Pos</th>\n",
       "      <th>Allele_ref</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>LoF_flags</th>\n",
       "      <th>filter</th>\n",
       "      <th>mu</th>\n",
       "      <th>AN</th>\n",
       "      <th>AC_unint</th>\n",
       "      <th>AN_nfe</th>\n",
       "      <th>AC_nfe_unint</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19789</th>\n",
       "      <td>CHEK2</td>\n",
       "      <td>ENSG00000183765</td>\n",
       "      <td>22</td>\n",
       "      <td>28687987.0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.041</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19790</th>\n",
       "      <td>CHEK2</td>\n",
       "      <td>ENSG00000183765</td>\n",
       "      <td>22</td>\n",
       "      <td>28687987.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.105</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19791</th>\n",
       "      <td>CHEK2</td>\n",
       "      <td>ENSG00000183765</td>\n",
       "      <td>22</td>\n",
       "      <td>28687987.0</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.041</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19792</th>\n",
       "      <td>CHEK2</td>\n",
       "      <td>ENSG00000183765</td>\n",
       "      <td>22</td>\n",
       "      <td>28687988.0</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19793</th>\n",
       "      <td>CHEK2</td>\n",
       "      <td>ENSG00000183765</td>\n",
       "      <td>22</td>\n",
       "      <td>28687988.0</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.094</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76699</th>\n",
       "      <td>TNRC6B</td>\n",
       "      <td>ENSG00000100354</td>\n",
       "      <td>22</td>\n",
       "      <td>40322852.0</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76700</th>\n",
       "      <td>TNRC6B</td>\n",
       "      <td>ENSG00000100354</td>\n",
       "      <td>22</td>\n",
       "      <td>40322852.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.139</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76701</th>\n",
       "      <td>TNRC6B</td>\n",
       "      <td>ENSG00000100354</td>\n",
       "      <td>22</td>\n",
       "      <td>40322852.0</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76702</th>\n",
       "      <td>TNRC6B</td>\n",
       "      <td>ENSG00000100354</td>\n",
       "      <td>22</td>\n",
       "      <td>40322853.0</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.073</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76703</th>\n",
       "      <td>TNRC6B</td>\n",
       "      <td>ENSG00000100354</td>\n",
       "      <td>22</td>\n",
       "      <td>40322853.0</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.041</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1024 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      GENE SYMBOL             Gene  Chrom         Pos Allele_ref Allele  \\\n",
       "19789       CHEK2  ENSG00000183765     22  28687987.0          C      A   \n",
       "19790       CHEK2  ENSG00000183765     22  28687987.0          C      T   \n",
       "19791       CHEK2  ENSG00000183765     22  28687987.0          C      G   \n",
       "19792       CHEK2  ENSG00000183765     22  28687988.0          T      G   \n",
       "19793       CHEK2  ENSG00000183765     22  28687988.0          T      C   \n",
       "...           ...              ...    ...         ...        ...    ...   \n",
       "76699      TNRC6B  ENSG00000100354     22  40322852.0          A      C   \n",
       "76700      TNRC6B  ENSG00000100354     22  40322852.0          A      G   \n",
       "76701      TNRC6B  ENSG00000100354     22  40322852.0          A      T   \n",
       "76702      TNRC6B  ENSG00000100354     22  40322853.0          G      C   \n",
       "76703      TNRC6B  ENSG00000100354     22  40322853.0          G      T   \n",
       "\n",
       "                   Consequence LoF_flags filter     mu  AN  AC_unint  AN_nfe  \\\n",
       "19789  splice_acceptor_variant         .    NaN  0.041 NaN       NaN     NaN   \n",
       "19790  splice_acceptor_variant         .    NaN  0.105 NaN       NaN     NaN   \n",
       "19791  splice_acceptor_variant         .    NaN  0.041 NaN       NaN     NaN   \n",
       "19792  splice_acceptor_variant         .    NaN  0.030 NaN       NaN     NaN   \n",
       "19793  splice_acceptor_variant         .    NaN  0.094 NaN       NaN     NaN   \n",
       "...                        ...       ...    ...    ...  ..       ...     ...   \n",
       "76699  splice_acceptor_variant         .    NaN  0.030 NaN       NaN     NaN   \n",
       "76700  splice_acceptor_variant         .    NaN  0.139 NaN       NaN     NaN   \n",
       "76701  splice_acceptor_variant         .    NaN  0.030 NaN       NaN     NaN   \n",
       "76702  splice_acceptor_variant         .    NaN  0.073 NaN       NaN     NaN   \n",
       "76703  splice_acceptor_variant         .    NaN  0.041 NaN       NaN     NaN   \n",
       "\n",
       "       AC_nfe_unint  \n",
       "19789           NaN  \n",
       "19790           NaN  \n",
       "19791           NaN  \n",
       "19792           NaN  \n",
       "19793           NaN  \n",
       "...             ...  \n",
       "76699           NaN  \n",
       "76700           NaN  \n",
       "76701           NaN  \n",
       "76702           NaN  \n",
       "76703           NaN  \n",
       "\n",
       "[1024 rows x 14 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "cecdc06d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_variants.to_csv(\"../data/variants_splice_acceptor_donor_AD_lof_strong.tsv\", sep = \"\\t\", index = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d07bec30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE SYMBOL</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Pos</th>\n",
       "      <th>Allele_ref</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>LoF_flags</th>\n",
       "      <th>filter</th>\n",
       "      <th>mu</th>\n",
       "      <th>AN</th>\n",
       "      <th>AC_unint</th>\n",
       "      <th>AN_nfe</th>\n",
       "      <th>AC_nfe_unint</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACVRL1</td>\n",
       "      <td>ENSG00000139567</td>\n",
       "      <td>12.0</td>\n",
       "      <td>51913097.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>NAGNAG_SITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.139</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACVRL1</td>\n",
       "      <td>ENSG00000139567</td>\n",
       "      <td>12.0</td>\n",
       "      <td>51913097.0</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>NAGNAG_SITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.020</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACVRL1</td>\n",
       "      <td>ENSG00000139567</td>\n",
       "      <td>12.0</td>\n",
       "      <td>51913098.0</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>NAGNAG_SITE</td>\n",
       "      <td>RF</td>\n",
       "      <td>0.020</td>\n",
       "      <td>232218.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>105398.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACVRL1</td>\n",
       "      <td>ENSG00000139567</td>\n",
       "      <td>12.0</td>\n",
       "      <td>51913098.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>NAGNAG_SITE</td>\n",
       "      <td>RF</td>\n",
       "      <td>0.094</td>\n",
       "      <td>232218.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>105398.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ACVRL1</td>\n",
       "      <td>ENSG00000139567</td>\n",
       "      <td>12.0</td>\n",
       "      <td>51913351.0</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>splice_donor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.051</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96717</th>\n",
       "      <td>ZNF462</td>\n",
       "      <td>ENSG00000148143</td>\n",
       "      <td>9.0</td>\n",
       "      <td>107009670.0</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>splice_donor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.020</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96718</th>\n",
       "      <td>ZNF462</td>\n",
       "      <td>ENSG00000148143</td>\n",
       "      <td>9.0</td>\n",
       "      <td>107010821.0</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.020</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96719</th>\n",
       "      <td>ZNF462</td>\n",
       "      <td>ENSG00000148143</td>\n",
       "      <td>9.0</td>\n",
       "      <td>107010821.0</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96720</th>\n",
       "      <td>ZNF462</td>\n",
       "      <td>ENSG00000148143</td>\n",
       "      <td>9.0</td>\n",
       "      <td>107010822.0</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.051</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96721</th>\n",
       "      <td>ZNF462</td>\n",
       "      <td>ENSG00000148143</td>\n",
       "      <td>9.0</td>\n",
       "      <td>107010822.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.094</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>31874 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      GENE SYMBOL             Gene  Chrom          Pos Allele_ref Allele  \\\n",
       "3          ACVRL1  ENSG00000139567   12.0   51913097.0          A      G   \n",
       "4          ACVRL1  ENSG00000139567   12.0   51913097.0          A      T   \n",
       "5          ACVRL1  ENSG00000139567   12.0   51913098.0          G      C   \n",
       "6          ACVRL1  ENSG00000139567   12.0   51913098.0          G      A   \n",
       "23         ACVRL1  ENSG00000139567   12.0   51913351.0          G      C   \n",
       "...           ...              ...    ...          ...        ...    ...   \n",
       "96717      ZNF462  ENSG00000148143    9.0  107009670.0          T      G   \n",
       "96718      ZNF462  ENSG00000148143    9.0  107010821.0          A      T   \n",
       "96719      ZNF462  ENSG00000148143    9.0  107010821.0          A      C   \n",
       "96720      ZNF462  ENSG00000148143    9.0  107010822.0          G      C   \n",
       "96721      ZNF462  ENSG00000148143    9.0  107010822.0          G      A   \n",
       "\n",
       "                   Consequence    LoF_flags filter     mu        AN  AC_unint  \\\n",
       "3      splice_acceptor_variant  NAGNAG_SITE    NaN  0.139       NaN       NaN   \n",
       "4      splice_acceptor_variant  NAGNAG_SITE    NaN  0.020       NaN       NaN   \n",
       "5      splice_acceptor_variant  NAGNAG_SITE     RF  0.020  232218.0       0.0   \n",
       "6      splice_acceptor_variant  NAGNAG_SITE     RF  0.094  232218.0       1.0   \n",
       "23        splice_donor_variant            .    NaN  0.051       NaN       NaN   \n",
       "...                        ...          ...    ...    ...       ...       ...   \n",
       "96717     splice_donor_variant            .    NaN  0.020       NaN       NaN   \n",
       "96718  splice_acceptor_variant            .    NaN  0.020       NaN       NaN   \n",
       "96719  splice_acceptor_variant            .    NaN  0.030       NaN       NaN   \n",
       "96720  splice_acceptor_variant            .    NaN  0.051       NaN       NaN   \n",
       "96721  splice_acceptor_variant            .    NaN  0.094       NaN       NaN   \n",
       "\n",
       "         AN_nfe  AC_nfe_unint  \n",
       "3           NaN           NaN  \n",
       "4           NaN           NaN  \n",
       "5      105398.0           0.0  \n",
       "6      105398.0           1.0  \n",
       "23          NaN           NaN  \n",
       "...         ...           ...  \n",
       "96717       NaN           NaN  \n",
       "96718       NaN           NaN  \n",
       "96719       NaN           NaN  \n",
       "96720       NaN           NaN  \n",
       "96721       NaN           NaN  \n",
       "\n",
       "[31874 rows x 14 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "37db0dee",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "telangiectasia, hereditary hemorrhagic, type 2, ADNP-related multiple congenital anomalies - intellectual disability - autism spectrum disorder, AHDC1-related intellectual disability - obstructive sleep apnea - mild dysmorphism syndrome, complex neurodevelopmental disorder, KBG syndrome, syndromic intellectual disability, gastric adenocarcinoma and proximal polyposis of the stomach, classic or attenuated familial adenomatous polyposis, hypercholesterolemia, autosomal dominant, type B, Coffin-Siris syndrome, Bohring-Opitz syndrome, hereditary breast carcinoma, myofibrillar myopathy, dilated cardiomyopathy, BAP1-related tumor predisposition syndrome, Dias-Logan syndrome, juvenile polyposis syndrome, pulmonary arterial hypertension, breast-ovarian cancer, familial, susceptibility to, 1, breast-ovarian cancer, familial, susceptibility to, 2, familial ovarian cancer, hyperparathyroidism 2 with jaw tumors, hereditary diffuse gastric adenocarcinoma, multiple endocrine neoplasia type 4, melanoma-pancreatic cancer syndrome, CHARGE syndrome, osteogenesis imperfecta type 2, osteogenesis imperfecta type 1, osteogenesis imperfecta type 3, spondyloepiphyseal dysplasia congenita, achondrogenesis type II, Kniest dysplasia, platyspondylic dysplasia, Torrance type, spondyloperipheral dysplasia, spondylometaphyseal dysplasia, Ehlers-Danlos syndrome, vascular type, Ehlers-Danlos syndrome, classic type, Rubinstein-Taybi syndrome, severe intellectual disability-progressive spastic diplegia syndrome, pseudohypoaldosteronism type 2E, Brooke-Spiegler syndrome, DDX41-related hematologic malignancy predisposition syndrome, pleuropulmonary blastoma, familial isolated arrhythmogenic right ventricular dysplasia, arrhythmogenic right ventricular cardiomyopathy, arrhythmogenic cardiomyopathy with woolly hair and keratoderma, mandibulofacial dysostosis-microcephaly syndrome, Kleefstra syndrome, hereditary hemorrhagic telangiectasia, Rubinstein-Taybi syndrome due to EP300 haploinsufficiency, craniosynostosis 4, thrombocytopenia 5, exostoses, multiple, type 1, exostoses, multiple, type 2, branchio-oto-renal syndrome, nonsyndromic genetic hearing loss, familial thoracic aortic aneurysm and aortic dissection, Marfan syndrome, Pfeiffer syndrome type 1, hereditary leiomyomatosis and renal cell cancer, Birt-Hogg-Dube syndrome, myofibrillar myopathy 5, anterior segment dysgenesis 3, FOXG1 disorder, intellectual disability-severe speech delay-mild dysmorphism syndrome, GATA2 deficiency with susceptibility to MDS/AML, hypoparathyroidism-deafness-renal disease syndrome, structural congenital heart disease, multiple types - GATA4, severe intellectual disability-poor language-strabismus-grimacing face-long fingers syndrome, renal cysts and diabetes syndrome, neurodevelopmental disorder-craniofacial dysmorphism-cardiac defect-hip dysplasia syndrome, Koolen-de Vries syndrome, KAT6B-related multiple congenital anomalies syndrome, long QT syndrome, short QT syndrome, neonatal-onset developmental and epileptic encephalopathy, microcephaly with or without chorioretinopathy, lymphedema, or intellectual disability, Wiedemann-Steiner syndrome, complex neurodevelopmental disorder with motor features, Kabuki syndrome 1, nail-patella syndrome, hereditary pheochromocytoma-paraganglioma, multiple endocrine neoplasia type 1, Waardenburg syndrome type 2, Lynch syndrome, hypertrophic cardiomyopathy, neurofibromatosis type 1, neurofibromatosis type 2, brain malformations with or without urinary tract defects, Marshall-Smith syndrome, Malan overgrowth syndrome, Cornelia de Lange syndrome, Bosch-Boonstra-Schaaf optic atrophy syndrome, autosomal dominant pseudohypoaldosteronism type 1, Sotos syndrome, focal segmental glomerulosclerosis 7, Waardenburg syndrome, PAX6-related ocular dysgenesis, congenital anomalies of kidney and urinary tract syndrome with or without hearing loss, abnormal ears, or developmental delay, PHIP-related behavioral problems-intellectual disability-obesity-dysmorphic features syndrome, autosomal dominant polycystic kidney disease, Charcot-Marie-Tooth disease type 1A, intellectual disability-microcephaly-strabismus-behavioral abnormalities syndrome, infantile convulsions and choreoathetosis, nevoid basal cell carcinoma syndrome, PTEN hamartoma tumor syndrome, Noonan syndrome, Noonan syndrome with multiple lentigines, Smith-Magenis syndrome, retinoblastoma, multiple endocrine neoplasia type 2B, multiple endocrine neoplasia type 2A, Diamond-Blackfan anemia, hereditary thrombocytopenia and hematologic cancer predisposition syndrome, SATB2 associated disorder, developmental and epileptic encephalopathy, generalized epilepsy with febrile seizures plus, Dravet syndrome, Brugada syndrome, familial long QT syndrome, Schinzel-Giedion syndrome, Phelan-McDermid syndrome, SIN3A-related intellectual disability syndrome, GLUT1 deficiency syndrome, aneurysm-osteoarthritis syndrome, generalized juvenile polyposis/juvenile polyposis coli, Myhre syndrome, juvenile polyposis/hereditary hemorrhagic telangiectasia syndrome, rhabdoid tumor predisposition syndrome 2, rhabdoid tumor predisposition syndrome 1, Waardenburg syndrome type 4C, Lamb-Shaffer syndrome, campomelic dysplasia, Legius syndrome, Peutz-Jeghers syndrome, Holt-Oram syndrome, TCF12-related craniosynostosis, developmental delay with variable intellectual impairment and behavioral abnormalities, Pitt-Hopkins syndrome, Treacher-Collins syndrome, multiple self-healing squamous epithelioma, Loeys-Dietz syndrome, Loeys-Dietz syndrome 2, Li-Fraumeni syndrome 1, tuberous sclerosis, Saethre-Chotzen syndrome, Angelman syndrome, von Hippel-Lindau disease, DeSanto-Shinawi syndrome, Wilms tumor 1, Denys-Drash syndrome, Mowat-Wilson syndrome, weiss-kruszka syndrome, "
     ]
    }
   ],
   "source": [
    "for x in list(df_AD_strong_lof_ensg[\"DISEASE LABEL\"].unique()):\n",
    "    print(x, end = \", \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a8b63544",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"['DISEASE LABEL', 'HAPLOINSUFFICIENCY'] not in index\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Input \u001b[0;32mIn [23]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mdf_variants\u001b[49m\u001b[43m[\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mGENE SYMBOL\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mDISEASE LABEL\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mHAPLOINSUFFICIENCY\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mdrop_duplicates()\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/py39/lib/python3.9/site-packages/pandas/core/frame.py:3511\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3509\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m is_iterator(key):\n\u001b[1;32m   3510\u001b[0m         key \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(key)\n\u001b[0;32m-> 3511\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_indexer_strict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcolumns\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m[\u001b[38;5;241m1\u001b[39m]\n\u001b[1;32m   3513\u001b[0m \u001b[38;5;66;03m# take() does not accept boolean indexers\u001b[39;00m\n\u001b[1;32m   3514\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(indexer, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdtype\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m) \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mbool\u001b[39m:\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/py39/lib/python3.9/site-packages/pandas/core/indexes/base.py:5782\u001b[0m, in \u001b[0;36mIndex._get_indexer_strict\u001b[0;34m(self, key, axis_name)\u001b[0m\n\u001b[1;32m   5779\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   5780\u001b[0m     keyarr, indexer, new_indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reindex_non_unique(keyarr)\n\u001b[0;32m-> 5782\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_raise_if_missing\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkeyarr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis_name\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   5784\u001b[0m keyarr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtake(indexer)\n\u001b[1;32m   5785\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(key, Index):\n\u001b[1;32m   5786\u001b[0m     \u001b[38;5;66;03m# GH 42790 - Preserve name from an Index\u001b[39;00m\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/py39/lib/python3.9/site-packages/pandas/core/indexes/base.py:5845\u001b[0m, in \u001b[0;36mIndex._raise_if_missing\u001b[0;34m(self, key, indexer, axis_name)\u001b[0m\n\u001b[1;32m   5842\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNone of [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mkey\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m] are in the [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00maxis_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m]\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   5844\u001b[0m not_found \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(ensure_index(key)[missing_mask\u001b[38;5;241m.\u001b[39mnonzero()[\u001b[38;5;241m0\u001b[39m]]\u001b[38;5;241m.\u001b[39munique())\n\u001b[0;32m-> 5845\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mnot_found\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m not in index\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mKeyError\u001b[0m: \"['DISEASE LABEL', 'HAPLOINSUFFICIENCY'] not in index\""
     ]
    }
   ],
   "source": [
    "df_variants[[\"GENE SYMBOL\", \"DISEASE LABEL\", \"HAPLOINSUFFICIENCY\"]].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b2e2fd5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "186"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_variants[\"GENE SYMBOL\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "7eafb4d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15981"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "882e3b84",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/wx/z2j6vh_s593cjs7xx71k0q780000gq/T/ipykernel_8592/1573240805.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_variants[\"mu_pergen\"] = df_variants[\"mu\"] * 1.015 * 10 **-7\n"
     ]
    }
   ],
   "source": [
    "df_variants[\"mu_pergen\"] = df_variants[\"mu\"] * 1.015 * 10 **-7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3948e24a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.00022304493049999993"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_variants[\"mu_pergen\"].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "884e949d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4483.401607731229"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1/df_variants[\"mu_pergen\"].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "4280a22d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "816.3444456299998"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "3.66*10**6 / 4483.401607731229\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "cab34373",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['NAGNAG_SITE', '.', 'NON_CAN_SPLICE'], dtype=object)"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_variants[\"LoF_flags\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "beda4dce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE SYMBOL</th>\n",
       "      <th>Gene</th>\n",
       "      <th>DISEASE LABEL</th>\n",
       "      <th>MOI</th>\n",
       "      <th>CLASSIFICATION</th>\n",
       "      <th>GCEP</th>\n",
       "      <th>HAPLOINSUFFICIENCY</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Pos</th>\n",
       "      <th>Allele_ref</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>LoF_flags</th>\n",
       "      <th>filter</th>\n",
       "      <th>mu</th>\n",
       "      <th>AN</th>\n",
       "      <th>AC_unint</th>\n",
       "      <th>AN_nfe</th>\n",
       "      <th>AC_nfe_unint</th>\n",
       "      <th>mu_pergen</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACVRL1</td>\n",
       "      <td>ENSG00000139567</td>\n",
       "      <td>telangiectasia, hereditary hemorrhagic, type 2</td>\n",
       "      <td>AD</td>\n",
       "      <td>Definitive</td>\n",
       "      <td>Hemostasis Thrombosis</td>\n",
       "      <td>Sufficient Evidence for Haploinsufficiency</td>\n",
       "      <td>12.0</td>\n",
       "      <td>51913097.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>NAGNAG_SITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.139</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.410850e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACVRL1</td>\n",
       "      <td>ENSG00000139567</td>\n",
       "      <td>telangiectasia, hereditary hemorrhagic, type 2</td>\n",
       "      <td>AD</td>\n",
       "      <td>Definitive</td>\n",
       "      <td>Hemostasis Thrombosis</td>\n",
       "      <td>Sufficient Evidence for Haploinsufficiency</td>\n",
       "      <td>12.0</td>\n",
       "      <td>51913097.0</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>NAGNAG_SITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.020</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.030000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACVRL1</td>\n",
       "      <td>ENSG00000139567</td>\n",
       "      <td>telangiectasia, hereditary hemorrhagic, type 2</td>\n",
       "      <td>AD</td>\n",
       "      <td>Definitive</td>\n",
       "      <td>Hemostasis Thrombosis</td>\n",
       "      <td>Sufficient Evidence for Haploinsufficiency</td>\n",
       "      <td>12.0</td>\n",
       "      <td>51913098.0</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>NAGNAG_SITE</td>\n",
       "      <td>RF</td>\n",
       "      <td>0.020</td>\n",
       "      <td>232218.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>105398.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.030000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACVRL1</td>\n",
       "      <td>ENSG00000139567</td>\n",
       "      <td>telangiectasia, hereditary hemorrhagic, type 2</td>\n",
       "      <td>AD</td>\n",
       "      <td>Definitive</td>\n",
       "      <td>Hemostasis Thrombosis</td>\n",
       "      <td>Sufficient Evidence for Haploinsufficiency</td>\n",
       "      <td>12.0</td>\n",
       "      <td>51913098.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>NAGNAG_SITE</td>\n",
       "      <td>RF</td>\n",
       "      <td>0.094</td>\n",
       "      <td>232218.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>105398.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.541000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ACVRL1</td>\n",
       "      <td>ENSG00000139567</td>\n",
       "      <td>telangiectasia, hereditary hemorrhagic, type 2</td>\n",
       "      <td>AD</td>\n",
       "      <td>Definitive</td>\n",
       "      <td>Hemostasis Thrombosis</td>\n",
       "      <td>Sufficient Evidence for Haploinsufficiency</td>\n",
       "      <td>12.0</td>\n",
       "      <td>51913557.0</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.020</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.030000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111537</th>\n",
       "      <td>ZNF462</td>\n",
       "      <td>ENSG00000148143</td>\n",
       "      <td>weiss-kruszka syndrome</td>\n",
       "      <td>AD</td>\n",
       "      <td>Definitive</td>\n",
       "      <td>Syndromic Disorders</td>\n",
       "      <td>Sufficient Evidence for Haploinsufficiency</td>\n",
       "      <td>9.0</td>\n",
       "      <td>107009544.0</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.020</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.030000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111547</th>\n",
       "      <td>ZNF462</td>\n",
       "      <td>ENSG00000148143</td>\n",
       "      <td>weiss-kruszka syndrome</td>\n",
       "      <td>AD</td>\n",
       "      <td>Definitive</td>\n",
       "      <td>Syndromic Disorders</td>\n",
       "      <td>Sufficient Evidence for Haploinsufficiency</td>\n",
       "      <td>9.0</td>\n",
       "      <td>107010821.0</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.020</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.030000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111548</th>\n",
       "      <td>ZNF462</td>\n",
       "      <td>ENSG00000148143</td>\n",
       "      <td>weiss-kruszka syndrome</td>\n",
       "      <td>AD</td>\n",
       "      <td>Definitive</td>\n",
       "      <td>Syndromic Disorders</td>\n",
       "      <td>Sufficient Evidence for Haploinsufficiency</td>\n",
       "      <td>9.0</td>\n",
       "      <td>107010821.0</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.045000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111549</th>\n",
       "      <td>ZNF462</td>\n",
       "      <td>ENSG00000148143</td>\n",
       "      <td>weiss-kruszka syndrome</td>\n",
       "      <td>AD</td>\n",
       "      <td>Definitive</td>\n",
       "      <td>Syndromic Disorders</td>\n",
       "      <td>Sufficient Evidence for Haploinsufficiency</td>\n",
       "      <td>9.0</td>\n",
       "      <td>107010822.0</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.051</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.176500e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111550</th>\n",
       "      <td>ZNF462</td>\n",
       "      <td>ENSG00000148143</td>\n",
       "      <td>weiss-kruszka syndrome</td>\n",
       "      <td>AD</td>\n",
       "      <td>Definitive</td>\n",
       "      <td>Syndromic Disorders</td>\n",
       "      <td>Sufficient Evidence for Haploinsufficiency</td>\n",
       "      <td>9.0</td>\n",
       "      <td>107010822.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.094</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.541000e-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19599 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       GENE SYMBOL             Gene  \\\n",
       "3           ACVRL1  ENSG00000139567   \n",
       "4           ACVRL1  ENSG00000139567   \n",
       "5           ACVRL1  ENSG00000139567   \n",
       "6           ACVRL1  ENSG00000139567   \n",
       "28          ACVRL1  ENSG00000139567   \n",
       "...            ...              ...   \n",
       "111537      ZNF462  ENSG00000148143   \n",
       "111547      ZNF462  ENSG00000148143   \n",
       "111548      ZNF462  ENSG00000148143   \n",
       "111549      ZNF462  ENSG00000148143   \n",
       "111550      ZNF462  ENSG00000148143   \n",
       "\n",
       "                                         DISEASE LABEL MOI CLASSIFICATION  \\\n",
       "3       telangiectasia, hereditary hemorrhagic, type 2  AD     Definitive   \n",
       "4       telangiectasia, hereditary hemorrhagic, type 2  AD     Definitive   \n",
       "5       telangiectasia, hereditary hemorrhagic, type 2  AD     Definitive   \n",
       "6       telangiectasia, hereditary hemorrhagic, type 2  AD     Definitive   \n",
       "28      telangiectasia, hereditary hemorrhagic, type 2  AD     Definitive   \n",
       "...                                                ...  ..            ...   \n",
       "111537                          weiss-kruszka syndrome  AD     Definitive   \n",
       "111547                          weiss-kruszka syndrome  AD     Definitive   \n",
       "111548                          weiss-kruszka syndrome  AD     Definitive   \n",
       "111549                          weiss-kruszka syndrome  AD     Definitive   \n",
       "111550                          weiss-kruszka syndrome  AD     Definitive   \n",
       "\n",
       "                         GCEP                          HAPLOINSUFFICIENCY  \\\n",
       "3       Hemostasis Thrombosis  Sufficient Evidence for Haploinsufficiency   \n",
       "4       Hemostasis Thrombosis  Sufficient Evidence for Haploinsufficiency   \n",
       "5       Hemostasis Thrombosis  Sufficient Evidence for Haploinsufficiency   \n",
       "6       Hemostasis Thrombosis  Sufficient Evidence for Haploinsufficiency   \n",
       "28      Hemostasis Thrombosis  Sufficient Evidence for Haploinsufficiency   \n",
       "...                       ...                                         ...   \n",
       "111537    Syndromic Disorders  Sufficient Evidence for Haploinsufficiency   \n",
       "111547    Syndromic Disorders  Sufficient Evidence for Haploinsufficiency   \n",
       "111548    Syndromic Disorders  Sufficient Evidence for Haploinsufficiency   \n",
       "111549    Syndromic Disorders  Sufficient Evidence for Haploinsufficiency   \n",
       "111550    Syndromic Disorders  Sufficient Evidence for Haploinsufficiency   \n",
       "\n",
       "        Chrom          Pos Allele_ref Allele              Consequence  \\\n",
       "3        12.0   51913097.0          A      G  splice_acceptor_variant   \n",
       "4        12.0   51913097.0          A      T  splice_acceptor_variant   \n",
       "5        12.0   51913098.0          G      C  splice_acceptor_variant   \n",
       "6        12.0   51913098.0          G      A  splice_acceptor_variant   \n",
       "28       12.0   51913557.0          A      T  splice_acceptor_variant   \n",
       "...       ...          ...        ...    ...                      ...   \n",
       "111537    9.0  107009544.0          G      T  splice_acceptor_variant   \n",
       "111547    9.0  107010821.0          A      T  splice_acceptor_variant   \n",
       "111548    9.0  107010821.0          A      C  splice_acceptor_variant   \n",
       "111549    9.0  107010822.0          G      C  splice_acceptor_variant   \n",
       "111550    9.0  107010822.0          G      A  splice_acceptor_variant   \n",
       "\n",
       "          LoF_flags filter     mu        AN  AC_unint    AN_nfe  AC_nfe_unint  \\\n",
       "3       NAGNAG_SITE    NaN  0.139       NaN       NaN       NaN           NaN   \n",
       "4       NAGNAG_SITE    NaN  0.020       NaN       NaN       NaN           NaN   \n",
       "5       NAGNAG_SITE     RF  0.020  232218.0       0.0  105398.0           0.0   \n",
       "6       NAGNAG_SITE     RF  0.094  232218.0       1.0  105398.0           1.0   \n",
       "28                .    NaN  0.020       NaN       NaN       NaN           NaN   \n",
       "...             ...    ...    ...       ...       ...       ...           ...   \n",
       "111537            .    NaN  0.020       NaN       NaN       NaN           NaN   \n",
       "111547            .    NaN  0.020       NaN       NaN       NaN           NaN   \n",
       "111548            .    NaN  0.030       NaN       NaN       NaN           NaN   \n",
       "111549            .    NaN  0.051       NaN       NaN       NaN           NaN   \n",
       "111550            .    NaN  0.094       NaN       NaN       NaN           NaN   \n",
       "\n",
       "           mu_pergen  \n",
       "3       1.410850e-08  \n",
       "4       2.030000e-09  \n",
       "5       2.030000e-09  \n",
       "6       9.541000e-09  \n",
       "28      2.030000e-09  \n",
       "...              ...  \n",
       "111537  2.030000e-09  \n",
       "111547  2.030000e-09  \n",
       "111548  3.045000e-09  \n",
       "111549  5.176500e-09  \n",
       "111550  9.541000e-09  \n",
       "\n",
       "[19599 rows x 20 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0239ca82",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
