{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "98f19562",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import dask.dataframe as dd\n",
    "from dask.distributed import Client\n",
    "import pandas as pd\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "\n",
    "pd_data_dir = \"/home/djl34/lab_pd/data\"\n",
    "aso_data_dir = \"/home/djl34/lab_pd/aso/data\"\n",
    "KL_data_dir = \"/home/djl34/lab_pd/kl/data\"\n",
    "scratch_dir = \"/n/scratch3/users/d/djl34\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "717f180c",
   "metadata": {},
   "source": [
    "## get a list of genes that are AD and lof"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "54d4404d",
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = aso_data_dir + \"/Clingen-Gene-Disease-Summary-2023-06-24.csv\"\n",
    "df = pd.read_csv(filename)\n",
    "\n",
    "filename = aso_data_dir + \"/Clingen-Dosage-Sensitivity-2023-06-25.csv\"\n",
    "df_ds = pd.read_csv(filename)\n",
    "df_ds_haploinsufficient = df_ds[df_ds[\"HAPLOINSUFFICIENCY\"].isin(['Sufficient Evidence for Haploinsufficiency', 'Emerging Evidence for Haploinsufficiency'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "faa76d26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Gene Associated with Autosomal Recessive Phenotype',\n",
       "       'Sufficient Evidence for Haploinsufficiency',\n",
       "       'No Evidence for Haploinsufficiency',\n",
       "       'Little Evidence for Haploinsufficiency',\n",
       "       'Emerging Evidence for Haploinsufficiency',\n",
       "       'Dosage Sensitivity Unlikely'], dtype=object)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ds[\"HAPLOINSUFFICIENCY\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e49176dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Disputed', 'Definitive', 'Moderate', 'Limited',\n",
       "       'No Known Disease Relationship', 'Strong', 'Refuted'], dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"CLASSIFICATION\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9006b4e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_AD = df[df[\"MOI\"] == \"AD\"]\n",
    "\n",
    "df_AD_strong = df_AD[df_AD[\"CLASSIFICATION\"].isin([\"Strong\", \"Definitive\"])]\n",
    "\n",
    "df_AD_strong_lof = df_AD_strong.merge(df_ds_haploinsufficient, on = \"GENE SYMBOL\", how = \"inner\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9ea76676",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list = list(df_AD_strong_lof[\"GENE SYMBOL\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "29eb5e7b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "191"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "cbd8afa0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ACVRL1',\n",
       " 'ADNP',\n",
       " 'AHDC1',\n",
       " 'ANK2',\n",
       " 'ANKRD11',\n",
       " 'ANKRD17',\n",
       " 'APC',\n",
       " 'APOB',\n",
       " 'ARID1A',\n",
       " 'ARID1B',\n",
       " 'ARID2',\n",
       " 'ASH1L',\n",
       " 'ASXL1',\n",
       " 'ASXL3',\n",
       " 'ATM',\n",
       " 'AUTS2',\n",
       " 'BAG3',\n",
       " 'BAP1',\n",
       " 'BARD1',\n",
       " 'BCL11A',\n",
       " 'BMPR1A',\n",
       " 'BMPR2',\n",
       " 'BPTF',\n",
       " 'BRCA1',\n",
       " 'BRCA2',\n",
       " 'BRIP1',\n",
       " 'CDC73',\n",
       " 'CDH1',\n",
       " 'CDK13',\n",
       " 'CDKN1B',\n",
       " 'CDKN2A',\n",
       " 'CHD2',\n",
       " 'CHD7',\n",
       " 'CHD8',\n",
       " 'CHEK2',\n",
       " 'COL1A1',\n",
       " 'COL2A1',\n",
       " 'COL3A1',\n",
       " 'COL5A1',\n",
       " 'CREBBP',\n",
       " 'CSNK2A1',\n",
       " 'CTCF',\n",
       " 'CTNNB1',\n",
       " 'CUL3',\n",
       " 'CYLD',\n",
       " 'DDX41',\n",
       " 'DICER1',\n",
       " 'DSC2',\n",
       " 'DSG2',\n",
       " 'DSP',\n",
       " 'DYRK1A',\n",
       " 'EFTUD2',\n",
       " 'EHMT1',\n",
       " 'ENG',\n",
       " 'EP300',\n",
       " 'ERF',\n",
       " 'ETV6',\n",
       " 'EXT1',\n",
       " 'EXT2',\n",
       " 'EYA1',\n",
       " 'EYA4',\n",
       " 'FBN1',\n",
       " 'FGFR1',\n",
       " 'FH',\n",
       " 'FLCN',\n",
       " 'FLNC',\n",
       " 'FOXC1',\n",
       " 'FOXG1',\n",
       " 'FOXP1',\n",
       " 'GATA2',\n",
       " 'GATA3',\n",
       " 'GATA4',\n",
       " 'GATAD2B',\n",
       " 'GRIN2A',\n",
       " 'GRIN2B',\n",
       " 'HNF1B',\n",
       " 'HNRNPK',\n",
       " 'KANSL1',\n",
       " 'KAT6A',\n",
       " 'KAT6B',\n",
       " 'KCNH2',\n",
       " 'KCNQ1',\n",
       " 'KCNQ2',\n",
       " 'KIF11',\n",
       " 'KMT2A',\n",
       " 'KMT2B',\n",
       " 'KMT2C',\n",
       " 'KMT2D',\n",
       " 'KMT2E',\n",
       " 'KMT5B',\n",
       " 'LMNA',\n",
       " 'LMX1B',\n",
       " 'MAX',\n",
       " 'MBD5',\n",
       " 'MED13L',\n",
       " 'MEF2C',\n",
       " 'MEIS2',\n",
       " 'MEN1',\n",
       " 'MITF',\n",
       " 'MLH1',\n",
       " 'MSH2',\n",
       " 'MSH6',\n",
       " 'MYBPC3',\n",
       " 'MYT1L',\n",
       " 'NAA15',\n",
       " 'NBEA',\n",
       " 'NF1',\n",
       " 'NF2',\n",
       " 'NFIA',\n",
       " 'NFIX',\n",
       " 'NIPBL',\n",
       " 'NR2F1',\n",
       " 'NR3C2',\n",
       " 'NR4A2',\n",
       " 'NRXN1',\n",
       " 'NSD1',\n",
       " 'PAX2',\n",
       " 'PAX3',\n",
       " 'PAX6',\n",
       " 'PBX1',\n",
       " 'PHF21A',\n",
       " 'PHIP',\n",
       " 'PKD1',\n",
       " 'PKD2',\n",
       " 'PKP2',\n",
       " 'PMP22',\n",
       " 'PMS2',\n",
       " 'POGZ',\n",
       " 'PRRT2',\n",
       " 'PTCH1',\n",
       " 'PTEN',\n",
       " 'PTPN11',\n",
       " 'PURA',\n",
       " 'RAD21',\n",
       " 'RAD51C',\n",
       " 'RAD51D',\n",
       " 'RAI1',\n",
       " 'RB1',\n",
       " 'RET',\n",
       " 'RPS19',\n",
       " 'RPS24',\n",
       " 'RUNX1',\n",
       " 'SATB2',\n",
       " 'SCN1A',\n",
       " 'SCN2A',\n",
       " 'SCN5A',\n",
       " 'SDHAF2',\n",
       " 'SDHB',\n",
       " 'SDHC',\n",
       " 'SDHD',\n",
       " 'SETBP1',\n",
       " 'SHANK2',\n",
       " 'SHANK3',\n",
       " 'SIN3A',\n",
       " 'SLC2A1',\n",
       " 'SMAD3',\n",
       " 'SMAD4',\n",
       " 'SMARCA4',\n",
       " 'SMARCB1',\n",
       " 'SOX10',\n",
       " 'SOX5',\n",
       " 'SOX9',\n",
       " 'SPRED1',\n",
       " 'STK11',\n",
       " 'STXBP1',\n",
       " 'SYNGAP1',\n",
       " 'TBL1XR1',\n",
       " 'TBR1',\n",
       " 'TBX4',\n",
       " 'TBX5',\n",
       " 'TCF12',\n",
       " 'TCF20',\n",
       " 'TCF4',\n",
       " 'TCOF1',\n",
       " 'TGFBR1',\n",
       " 'TGFBR2',\n",
       " 'TMEM127',\n",
       " 'TNRC6B',\n",
       " 'TP53',\n",
       " 'TRIO',\n",
       " 'TRIP12',\n",
       " 'TSC1',\n",
       " 'TSC2',\n",
       " 'TTN',\n",
       " 'TWIST1',\n",
       " 'UBE3A',\n",
       " 'VHL',\n",
       " 'WAC',\n",
       " 'WT1',\n",
       " 'ZEB2',\n",
       " 'ZNF462']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c4738b24",
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = pd_data_dir + \"/biomart/ENSG_Genename_syn.tsv\"\n",
    "df_ensg = pd.read_csv(filename, sep = \"\\t\")\n",
    "df_ensg = df_ensg.rename({\"Gene name\": \"GENE SYMBOL\", \"Gene stable ID\" : \"Gene\"}, axis = 1)\n",
    "df_ensg = df_ensg[[\"GENE SYMBOL\", \"Gene\", \"Chromosome/scaffold name\"]].drop_duplicates()\n",
    "df_ensg = df_ensg[df_ensg[\"Chromosome/scaffold name\"].str.contains(\"CHR\") == False]\n",
    "df_ensg = df_ensg.rename({\"Gene name\": \"GENE SYMBOL\", \"Gene stable ID\" : \"Gene\"}, axis = 1)\n",
    "\n",
    "df_AD_strong_lof_ensg = df_AD_strong_lof.merge(df_ensg, on = \"GENE SYMBOL\", how = \"left\")\n",
    "\n",
    "ensg_gene_list = list(df_AD_strong_lof_ensg[\"Gene\"].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e18b8b7",
   "metadata": {},
   "source": [
    "## combine with variant data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c1f05a99",
   "metadata": {},
   "outputs": [],
   "source": [
    "client = Client()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "92f00f56",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom = str(22)\n",
    "\n",
    "variants = os.path.join(aso_data_dir, \"whole_gene/spliceai/\"+ chrom +\"/_metadata\")\n",
    "variants = dd.read_parquet(\"/\".join(variants.split(\"/\")[:-1]) + \"/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e7cfb8ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>Pos</th>\n",
       "      <th>Allele_ref</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature</th>\n",
       "      <th>MaxEntScan_alt</th>\n",
       "      <th>MaxEntScan_diff</th>\n",
       "      <th>MaxEntScan_ref</th>\n",
       "      <th>LaBranchoR_start</th>\n",
       "      <th>LaBranchoR_score</th>\n",
       "      <th>spliceai_gene</th>\n",
       "      <th>DS</th>\n",
       "      <th>DS_AG</th>\n",
       "      <th>DS_AL</th>\n",
       "      <th>DS_DG</th>\n",
       "      <th>DS_DL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>10736171</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>non_coding_transcript_exon_variant</td>\n",
       "      <td>ENSG00000277248</td>\n",
       "      <td>ENST00000615943</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>10736171</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>non_coding_transcript_exon_variant</td>\n",
       "      <td>ENSG00000277248</td>\n",
       "      <td>ENST00000615943</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>10736171</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>non_coding_transcript_exon_variant</td>\n",
       "      <td>ENSG00000277248</td>\n",
       "      <td>ENST00000615943</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>10736172</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>non_coding_transcript_exon_variant</td>\n",
       "      <td>ENSG00000277248</td>\n",
       "      <td>ENST00000615943</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>10736172</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>non_coding_transcript_exon_variant</td>\n",
       "      <td>ENSG00000277248</td>\n",
       "      <td>ENST00000615943</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM       Pos Allele_ref Allele                         Consequence  \\\n",
       "0     22  10736171          T      A  non_coding_transcript_exon_variant   \n",
       "1     22  10736171          T      C  non_coding_transcript_exon_variant   \n",
       "2     22  10736171          T      G  non_coding_transcript_exon_variant   \n",
       "3     22  10736172          T      A  non_coding_transcript_exon_variant   \n",
       "4     22  10736172          T      C  non_coding_transcript_exon_variant   \n",
       "\n",
       "              Gene          Feature MaxEntScan_alt MaxEntScan_diff  \\\n",
       "0  ENSG00000277248  ENST00000615943              .               .   \n",
       "1  ENSG00000277248  ENST00000615943              .               .   \n",
       "2  ENSG00000277248  ENST00000615943              .               .   \n",
       "3  ENSG00000277248  ENST00000615943              .               .   \n",
       "4  ENSG00000277248  ENST00000615943              .               .   \n",
       "\n",
       "  MaxEntScan_ref  LaBranchoR_start  LaBranchoR_score spliceai_gene   DS  \\\n",
       "0              .               NaN               NaN                0.0   \n",
       "1              .               NaN               NaN                0.0   \n",
       "2              .               NaN               NaN                0.0   \n",
       "3              .               NaN               NaN                0.0   \n",
       "4              .               NaN               NaN                0.0   \n",
       "\n",
       "   DS_AG  DS_AL  DS_DG  DS_DL  \n",
       "0    0.0    0.0    0.0    0.0  \n",
       "1    0.0    0.0    0.0    0.0  \n",
       "2    0.0    0.0    0.0    0.0  \n",
       "3    0.0    0.0    0.0    0.0  \n",
       "4    0.0    0.0    0.0    0.0  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "bdce6b39",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants_ad = variants[variants[\"Gene\"].isin(ensg_gene_list)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "681bc4dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "## we get a list of variants, in a certain chromosome that are ad strong genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0529d871",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants_ad = variants_ad.compute()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "32a0ac5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(list(variants_ad[\"Gene\"].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ecebf3d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>Pos</th>\n",
       "      <th>Allele_ref</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature</th>\n",
       "      <th>MaxEntScan_alt</th>\n",
       "      <th>MaxEntScan_diff</th>\n",
       "      <th>MaxEntScan_ref</th>\n",
       "      <th>LaBranchoR_start</th>\n",
       "      <th>LaBranchoR_score</th>\n",
       "      <th>spliceai_gene</th>\n",
       "      <th>DS</th>\n",
       "      <th>DS_AG</th>\n",
       "      <th>DS_AL</th>\n",
       "      <th>DS_DG</th>\n",
       "      <th>DS_DL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3513549</th>\n",
       "      <td>22</td>\n",
       "      <td>28687743</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>ENSG00000183765</td>\n",
       "      <td>ENST00000404276</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3513550</th>\n",
       "      <td>22</td>\n",
       "      <td>28687743</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>ENSG00000183765</td>\n",
       "      <td>ENST00000404276</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3513551</th>\n",
       "      <td>22</td>\n",
       "      <td>28687743</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>ENSG00000183765</td>\n",
       "      <td>ENST00000404276</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3513552</th>\n",
       "      <td>22</td>\n",
       "      <td>28687744</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>ENSG00000183765</td>\n",
       "      <td>ENST00000404276</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3513553</th>\n",
       "      <td>22</td>\n",
       "      <td>28687744</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>ENSG00000183765</td>\n",
       "      <td>ENST00000404276</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369931</th>\n",
       "      <td>22</td>\n",
       "      <td>23829501</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>ENSG00000099956</td>\n",
       "      <td>ENST00000644036</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SMARCB1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369932</th>\n",
       "      <td>22</td>\n",
       "      <td>23829501</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>ENSG00000099956</td>\n",
       "      <td>ENST00000644036</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SMARCB1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369933</th>\n",
       "      <td>22</td>\n",
       "      <td>23829502</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>ENSG00000099956</td>\n",
       "      <td>ENST00000644036</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SMARCB1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369934</th>\n",
       "      <td>22</td>\n",
       "      <td>23829502</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>ENSG00000099956</td>\n",
       "      <td>ENST00000644036</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SMARCB1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369935</th>\n",
       "      <td>22</td>\n",
       "      <td>23829502</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>ENSG00000099956</td>\n",
       "      <td>ENST00000644036</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SMARCB1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1952577 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         CHROM       Pos Allele_ref Allele          Consequence  \\\n",
       "3513549     22  28687743          C      A  3_prime_UTR_variant   \n",
       "3513550     22  28687743          C      G  3_prime_UTR_variant   \n",
       "3513551     22  28687743          C      T  3_prime_UTR_variant   \n",
       "3513552     22  28687744          C      A  3_prime_UTR_variant   \n",
       "3513553     22  28687744          C      G  3_prime_UTR_variant   \n",
       "...        ...       ...        ...    ...                  ...   \n",
       "369931      22  23829501          C      G       intron_variant   \n",
       "369932      22  23829501          C      T       intron_variant   \n",
       "369933      22  23829502          C      A       intron_variant   \n",
       "369934      22  23829502          C      G       intron_variant   \n",
       "369935      22  23829502          C      T       intron_variant   \n",
       "\n",
       "                    Gene          Feature MaxEntScan_alt MaxEntScan_diff  \\\n",
       "3513549  ENSG00000183765  ENST00000404276              .               .   \n",
       "3513550  ENSG00000183765  ENST00000404276              .               .   \n",
       "3513551  ENSG00000183765  ENST00000404276              .               .   \n",
       "3513552  ENSG00000183765  ENST00000404276              .               .   \n",
       "3513553  ENSG00000183765  ENST00000404276              .               .   \n",
       "...                  ...              ...            ...             ...   \n",
       "369931   ENSG00000099956  ENST00000644036              .               .   \n",
       "369932   ENSG00000099956  ENST00000644036              .               .   \n",
       "369933   ENSG00000099956  ENST00000644036              .               .   \n",
       "369934   ENSG00000099956  ENST00000644036              .               .   \n",
       "369935   ENSG00000099956  ENST00000644036              .               .   \n",
       "\n",
       "        MaxEntScan_ref  LaBranchoR_start  LaBranchoR_score spliceai_gene   DS  \\\n",
       "3513549              .               NaN               NaN                0.0   \n",
       "3513550              .               NaN               NaN                0.0   \n",
       "3513551              .               NaN               NaN                0.0   \n",
       "3513552              .               NaN               NaN                0.0   \n",
       "3513553              .               NaN               NaN                0.0   \n",
       "...                ...               ...               ...           ...  ...   \n",
       "369931               .               NaN               NaN       SMARCB1  0.0   \n",
       "369932               .               NaN               NaN       SMARCB1  0.0   \n",
       "369933               .               NaN               NaN       SMARCB1  0.0   \n",
       "369934               .               NaN               NaN       SMARCB1  0.0   \n",
       "369935               .               NaN               NaN       SMARCB1  0.0   \n",
       "\n",
       "         DS_AG  DS_AL  DS_DG  DS_DL  \n",
       "3513549    0.0    0.0    0.0    0.0  \n",
       "3513550    0.0    0.0    0.0    0.0  \n",
       "3513551    0.0    0.0    0.0    0.0  \n",
       "3513552    0.0    0.0    0.0    0.0  \n",
       "3513553    0.0    0.0    0.0    0.0  \n",
       "...        ...    ...    ...    ...  \n",
       "369931     0.0    0.0    0.0    0.0  \n",
       "369932     0.0    0.0    0.0    0.0  \n",
       "369933     0.0    0.0    0.0    0.0  \n",
       "369934     0.0    0.0    0.0    0.0  \n",
       "369935     0.0    0.0    0.0    0.0  \n",
       "\n",
       "[1952577 rows x 18 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants_ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "ec0adb04",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "684"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(variants_ad[variants_ad[\"Consequence\"] == \"splice_donor_variant\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "4061d2e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DS_DL\n",
       "1.00    166\n",
       "0.99     99\n",
       "0.98     52\n",
       "0.00     48\n",
       "1.98     48\n",
       "2.00     48\n",
       "0.97     31\n",
       "0.96     19\n",
       "0.90     18\n",
       "1.96     15\n",
       "0.93     10\n",
       "0.94     10\n",
       "1.92      9\n",
       "0.86      9\n",
       "1.76      7\n",
       "0.76      7\n",
       "1.70      6\n",
       "0.41      6\n",
       "0.50      6\n",
       "0.66      6\n",
       "0.95      5\n",
       "0.85      5\n",
       "1.94      4\n",
       "0.89      4\n",
       "0.16      3\n",
       "0.56      3\n",
       "0.55      3\n",
       "1.90      3\n",
       "1.86      2\n",
       "0.92      2\n",
       "0.82      2\n",
       "0.74      2\n",
       "0.84      2\n",
       "0.28      1\n",
       "0.15      1\n",
       "0.09      1\n",
       "0.29      1\n",
       "0.30      1\n",
       "0.37      1\n",
       "0.47      1\n",
       "0.52      1\n",
       "0.54      1\n",
       "0.61      1\n",
       "1.44      1\n",
       "0.13      1\n",
       "0.65      1\n",
       "0.67      1\n",
       "0.73      1\n",
       "0.78      1\n",
       "0.79      1\n",
       "0.80      1\n",
       "0.91      1\n",
       "0.06      1\n",
       "0.83      1\n",
       "0.88      1\n",
       "0.87      1\n",
       "0.18      1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants_ad[variants_ad[\"Consequence\"] == \"splice_donor_variant\"].value_counts(\"DS_DL\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "58e661ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "696"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(variants_ad[variants_ad[\"Consequence\"] == \"splice_acceptor_variant\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9701ba15",
   "metadata": {},
   "source": [
    "## which of these variants are gain of misplicing?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ac3d549",
   "metadata": {},
   "source": [
    "spliceAI donor/ acceptor gain score at a non-canonical site ≥ 0.1 AND (ii) MaxEntScan donor/acceptor score with the ALT allele at the site ≥ 2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "c94f8d85",
   "metadata": {},
   "outputs": [],
   "source": [
    "misplicing_gain = variants_ad[(variants_ad[\"DS_AG\"] > 0.1) | (variants_ad[\"DS_DG\"] > 0.1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "970a1965",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_27022/706118262.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  misplicing_gain[\"MaxEntScan_alt\"] = misplicing_gain[\"MaxEntScan_alt\"].replace(\".\", None)\n",
      "/tmp/ipykernel_27022/706118262.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  misplicing_gain[\"MaxEntScan_alt\"] = misplicing_gain[\"MaxEntScan_alt\"].astype(float)\n",
      "/tmp/ipykernel_27022/706118262.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  misplicing_gain[\"MaxEntScan_ref\"] = misplicing_gain[\"MaxEntScan_ref\"].replace(\".\", None)\n",
      "/tmp/ipykernel_27022/706118262.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  misplicing_gain[\"MaxEntScan_ref\"] = misplicing_gain[\"MaxEntScan_ref\"].astype(float)\n"
     ]
    }
   ],
   "source": [
    "misplicing_gain[\"MaxEntScan_alt\"] = misplicing_gain[\"MaxEntScan_alt\"].replace(\".\", None)\n",
    "misplicing_gain[\"MaxEntScan_alt\"] = misplicing_gain[\"MaxEntScan_alt\"].astype(float)\n",
    "\n",
    "misplicing_gain[\"MaxEntScan_ref\"] = misplicing_gain[\"MaxEntScan_ref\"].replace(\".\", None)\n",
    "misplicing_gain[\"MaxEntScan_ref\"] = misplicing_gain[\"MaxEntScan_ref\"].astype(float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "b79cb08e",
   "metadata": {},
   "outputs": [],
   "source": [
    "misplicing_gain = misplicing_gain[misplicing_gain[\"MaxEntScan_alt\"] > 2]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ac10091",
   "metadata": {},
   "source": [
    "need to filter out severe splice destroying variants: SpliceAI donor/acceptor loss score ≥ 0.1 at a canonical splice site) AND (MaxEntScan donor/ acceptor score with the ALT allele < MaxEntScan donor/acceptor score with the REF allele) AND [(MaxEntScan donor/acceptor score with the ALT allele < 2) OR (MaxEntScan donor/acceptor score with the ALT allele < 0.3 × MaxEntScan donor/acceptor score with the REF allele)]."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "b5b5816d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## remove non-severe\n",
    "\n",
    "misplicing_gain = misplicing_gain[(((misplicing_gain[\"DS_AL\"] > 0.1) |(misplicing_gain[\"DS_DL\"] > 0.1)) & \n",
    "                (misplicing_gain[\"MaxEntScan_alt\"] < misplicing_gain[\"MaxEntScan_ref\"]) & \n",
    "                ((misplicing_gain[\"MaxEntScan_alt\"] < 2) | \n",
    "                 (misplicing_gain[\"MaxEntScan_alt\"] < 0.3 * misplicing_gain[\"MaxEntScan_ref\"]))) == False]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "cdf8f86b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['splice_polypyrimidine_tract_variant&intron_variant',\n",
       "       'intron_variant',\n",
       "       'splice_region_variant&splice_polypyrimidine_tract_variant&intron_variant',\n",
       "       'splice_acceptor_variant',\n",
       "       'splice_donor_region_variant&intron_variant',\n",
       "       'splice_donor_5th_base_variant&intron_variant',\n",
       "       'splice_donor_variant'], dtype=object)"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "misplicing_gain[\"Consequence\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "864d8190",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Consequence\n",
       "splice_polypyrimidine_tract_variant&intron_variant                          242\n",
       "splice_region_variant&splice_polypyrimidine_tract_variant&intron_variant    211\n",
       "splice_donor_region_variant&intron_variant                                  163\n",
       "intron_variant                                                               82\n",
       "splice_donor_5th_base_variant&intron_variant                                 52\n",
       "splice_acceptor_variant                                                      32\n",
       "splice_donor_variant                                                          5\n",
       "dtype: int64"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "misplicing_gain.value_counts(\"Consequence\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "c19c1a8e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "787"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(misplicing_gain)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b6e7a54",
   "metadata": {},
   "source": [
    "## which of these are exon skipping or intron retention?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25c65929",
   "metadata": {},
   "source": [
    "(2) Exon skipping or intron retention (skipping or retention): SpliceAI donor/acceptor loss score at any canonical site ≥ 0.1 without an accompanying gain of mis-splicing by SpliceAI (donor/acceptor gain score < 0.1 at any non-canonical splice site)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "8da5b884",
   "metadata": {},
   "outputs": [],
   "source": [
    "exon_skip_intron_ret = variants_ad[(variants_ad[\"DS_AL\"] > 0.1) | (variants_ad[\"DS_DL\"] > 0.1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "eef32808",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>Pos</th>\n",
       "      <th>Allele_ref</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature</th>\n",
       "      <th>MaxEntScan_alt</th>\n",
       "      <th>MaxEntScan_diff</th>\n",
       "      <th>MaxEntScan_ref</th>\n",
       "      <th>LaBranchoR_start</th>\n",
       "      <th>LaBranchoR_score</th>\n",
       "      <th>spliceai_gene</th>\n",
       "      <th>DS</th>\n",
       "      <th>DS_AG</th>\n",
       "      <th>DS_AL</th>\n",
       "      <th>DS_DG</th>\n",
       "      <th>DS_DL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3518533</th>\n",
       "      <td>22</td>\n",
       "      <td>28689404</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>ENSG00000183765</td>\n",
       "      <td>ENST00000404276</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEK2CHEK2</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3518616</th>\n",
       "      <td>22</td>\n",
       "      <td>28689432</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>ENSG00000183765</td>\n",
       "      <td>ENST00000404276</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEK2CHEK2</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3518617</th>\n",
       "      <td>22</td>\n",
       "      <td>28689432</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>ENSG00000183765</td>\n",
       "      <td>ENST00000404276</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEK2CHEK2</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3518618</th>\n",
       "      <td>22</td>\n",
       "      <td>28689432</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>ENSG00000183765</td>\n",
       "      <td>ENST00000404276</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEK2CHEK2</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3518619</th>\n",
       "      <td>22</td>\n",
       "      <td>28689433</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>ENSG00000183765</td>\n",
       "      <td>ENST00000404276</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEK2CHEK2</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357676</th>\n",
       "      <td>22</td>\n",
       "      <td>23825416</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>splice_donor_variant</td>\n",
       "      <td>ENSG00000099956</td>\n",
       "      <td>ENST00000644036</td>\n",
       "      <td>1.513</td>\n",
       "      <td>8.273</td>\n",
       "      <td>9.786</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SMARCB1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357677</th>\n",
       "      <td>22</td>\n",
       "      <td>23825416</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>splice_donor_variant</td>\n",
       "      <td>ENSG00000099956</td>\n",
       "      <td>ENST00000644036</td>\n",
       "      <td>1.282</td>\n",
       "      <td>8.504</td>\n",
       "      <td>9.786</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SMARCB1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.32</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357678</th>\n",
       "      <td>22</td>\n",
       "      <td>23825417</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>splice_donor_variant</td>\n",
       "      <td>ENSG00000099956</td>\n",
       "      <td>ENST00000644036</td>\n",
       "      <td>1.602</td>\n",
       "      <td>8.183</td>\n",
       "      <td>9.786</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SMARCB1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357679</th>\n",
       "      <td>22</td>\n",
       "      <td>23825417</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>splice_donor_variant</td>\n",
       "      <td>ENSG00000099956</td>\n",
       "      <td>ENST00000644036</td>\n",
       "      <td>2.032</td>\n",
       "      <td>7.754</td>\n",
       "      <td>9.786</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SMARCB1</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357680</th>\n",
       "      <td>22</td>\n",
       "      <td>23825417</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>splice_donor_variant</td>\n",
       "      <td>ENSG00000099956</td>\n",
       "      <td>ENST00000644036</td>\n",
       "      <td>2.139</td>\n",
       "      <td>7.647</td>\n",
       "      <td>9.786</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SMARCB1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11719 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         CHROM       Pos Allele_ref Allele           Consequence  \\\n",
       "3518533     22  28689404          G      C        intron_variant   \n",
       "3518616     22  28689432          A      C        intron_variant   \n",
       "3518617     22  28689432          A      G        intron_variant   \n",
       "3518618     22  28689432          A      T        intron_variant   \n",
       "3518619     22  28689433          C      A        intron_variant   \n",
       "...        ...       ...        ...    ...                   ...   \n",
       "357676      22  23825416          G      C  splice_donor_variant   \n",
       "357677      22  23825416          G      T  splice_donor_variant   \n",
       "357678      22  23825417          T      A  splice_donor_variant   \n",
       "357679      22  23825417          T      C  splice_donor_variant   \n",
       "357680      22  23825417          T      G  splice_donor_variant   \n",
       "\n",
       "                    Gene          Feature MaxEntScan_alt MaxEntScan_diff  \\\n",
       "3518533  ENSG00000183765  ENST00000404276              .               .   \n",
       "3518616  ENSG00000183765  ENST00000404276              .               .   \n",
       "3518617  ENSG00000183765  ENST00000404276              .               .   \n",
       "3518618  ENSG00000183765  ENST00000404276              .               .   \n",
       "3518619  ENSG00000183765  ENST00000404276              .               .   \n",
       "...                  ...              ...            ...             ...   \n",
       "357676   ENSG00000099956  ENST00000644036          1.513           8.273   \n",
       "357677   ENSG00000099956  ENST00000644036          1.282           8.504   \n",
       "357678   ENSG00000099956  ENST00000644036          1.602           8.183   \n",
       "357679   ENSG00000099956  ENST00000644036          2.032           7.754   \n",
       "357680   ENSG00000099956  ENST00000644036          2.139           7.647   \n",
       "\n",
       "        MaxEntScan_ref  LaBranchoR_start  LaBranchoR_score spliceai_gene  \\\n",
       "3518533              .               NaN               NaN    CHEK2CHEK2   \n",
       "3518616              .               NaN               NaN    CHEK2CHEK2   \n",
       "3518617              .               NaN               NaN    CHEK2CHEK2   \n",
       "3518618              .               NaN               NaN    CHEK2CHEK2   \n",
       "3518619              .               NaN               NaN    CHEK2CHEK2   \n",
       "...                ...               ...               ...           ...   \n",
       "357676           9.786               NaN               NaN       SMARCB1   \n",
       "357677           9.786               NaN               NaN       SMARCB1   \n",
       "357678           9.786               NaN               NaN       SMARCB1   \n",
       "357679           9.786               NaN               NaN       SMARCB1   \n",
       "357680           9.786               NaN               NaN       SMARCB1   \n",
       "\n",
       "           DS  DS_AG  DS_AL  DS_DG  DS_DL  \n",
       "3518533  0.28    0.0    0.0   0.28   0.14  \n",
       "3518616  0.16    0.0    0.0   0.06   0.16  \n",
       "3518617  0.16    0.0    0.0   0.14   0.16  \n",
       "3518618  0.16    0.0    0.0   0.12   0.16  \n",
       "3518619  0.16    0.0    0.0   0.16   0.16  \n",
       "...       ...    ...    ...    ...    ...  \n",
       "357676   1.00    0.0    0.0   0.01   1.00  \n",
       "357677   1.00    0.0    0.0   0.32   1.00  \n",
       "357678   1.00    0.0    0.0   0.01   1.00  \n",
       "357679   0.56    0.0    0.0   0.01   0.56  \n",
       "357680   1.00    0.0    0.0   0.00   1.00  \n",
       "\n",
       "[11719 rows x 18 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_skip_intron_ret"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c86aefe",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3bfa0905",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "665497a3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
